quickjs.c

```diff
@@ -5389,13 +5482,22 @@ static void mark_children(JSRuntime *rt, JSGCObjectHeader *gp,
     case JS_GC_OBJ_TYPE_FUNCTION_BYTECODE:
         /* the template objects can be part of a cycle */
         {
+            InlineCacheRingItem *buffer;
             JSFunctionBytecode *b = (JSFunctionBytecode *)gp;
-            int i;
+            int i, j;
             for(i = 0; i < b->cpool_count; i++) {
                 JS_MarkValue(rt, b->cpool[i], mark_func);
             }
             if (b->realm)
                 mark_func(rt, &b->realm->header);
+            if (b->ic) {
+                for (i = 0; i < b->ic->count; i++) {
+                    buffer = b->ic->cache[i].buffer;
+                    for (j = 0; j < IC_CACHE_ITEM_CAPACITY; j++)
```

Comment:
```suggestion
                    for (j = 0; j < countof(b->ic->cache[0].buffer); j++)
```
Admittedly clunky-looking because it repeats the expression from the preceding line. Another way to write it:
```c
InlineCacheRingItem *buffer, (*buffers)[IC_CACHE_ITEM_CAPACITY];
// ...
if (b->ic) {
    for (i = 0; i < b->ic->count; i++) {
        buffers = &b->ic->cache[i].buffer;
        for (buffer = *buffers; buffer != *buffers + countof(*buffers); buffer++)
            if (buffer->shape)
                mark_func(rt, &buffer->shape->header);
    }
}
```
Encoding the array size in the type makes it harder to overrun (but not impossible, of course - this is C, after all.)

And a final improvement is to add an endof() macro:
```c
#define endof(x) ((x) + countof(x))
```
And replace ` *buffers + countof(*buffers)` with `endof(*buffers)`.